onnx/backend/test/case/node/negativeloglikelihoodloss.py

```diff
@@ -21,16 +21,31 @@ def compute_negative_log_likelihood_loss(input, target, weight=None, reduction='
         for i in range(N):
             if target[i] != ignore_index:
                 neg_gather_element_input[i] = -input[i][target[i]]
-    else:
+    elif len(input_shape) == 3:
+        N, C, dim1 = input_shape
+        neg_gather_element_input = np.zeros((N, dim1), dtype=np.float32)
+        for i in range(N):
+            for d1 in range(dim1):
+                if target[i][d1] != ignore_index:
+                    neg_gather_element_input[i][d1] = -input[i][target[i][d1]][d1]
+    elif len(input_shape) == 4:
         N, C, dim1, dim2 = input_shape
         neg_gather_element_input = np.zeros((N, dim1, dim2), dtype=np.float32)
         for i in range(N):
             for d1 in range(dim1):
                 for d2 in range(dim2):
                     if target[i][d1][d2] != ignore_index:
                         neg_gather_element_input[i][d1][d2] = -input[i][target[i][d1][d2]][d1][d2]
+    else:
+        raise NotImplementedError
 
     loss = neg_gather_element_input
+
+    if weight is None and ignore_index is not None:
+        c = input_shape[1]
+        weight = np.ones(c, dtype=np.float32)
+        weight[ignore_index] = np.float32(0)  # type: ignore
```

Comment:
I am a bit confused: it looks like "ignore_index" is being handled in two places (both in the conditions above which update neg_gather_element_input only for values other than ignore_index, and here using weight). Is that redundant?

Reply:
it's not really redundant. 
For example, in case the reduction is mean and ignore_index is specified, we don't want to include all the target values in the mean operation but only the non-ignored ones. This is why this if is added.

Reply:
Yes, but I think this makes the earlier "if-condition" in line 22/29/37 redundant, I think. IIUC, if ignore_index is specified, then we end up introducing a weight and ensure that the weight is zero for ignore_index, and end up executing line 72 "loss = gather_weight * loss". So, why do we need the earlier if-condition in line 22 etc.? (I might be wrong. Please verify before changing it.)